.github/workflows/ci.yaml

```diff
@@ -243,11 +232,12 @@ jobs:
       - uses: extractions/setup-just@v1
         with:
           just-version: "1.23.0"
-      - uses: actions/cache@v4
-        name: glaredb cache
+      - uses: actions/cache/restore@v4
+        name: build cache
         with:
-          path: target/debug/glaredb
-          key: ${{ github.run_id }}
+          path: |
+            target/
+          key: ${{ runner.os }}-cargo--target-${{github.sha}}
```

Comment:
what is the purpose of this new cache? Hard time telling from the changes.

Reply:
There's sort of a stacked cache (looking at the build cache is really where this might become clear):
- cache toolchains based on the `rust-toolchain` file (restores before, saves after the build task. (changes rarely, used between different branches so we're not downloading big files
- cache the user's `.cargo` directory based on the `Cargo.lock. (restores before and saves after the build task.) should prevent us from re-downloading dependencies.
- cache the workspace except for the glaredb binary (restores before and after the build task _only_, keyed off of the Cargo.lock, hopefully with the effect of saving as much incremental build time as possible in a fairly stable way and preventing rebuilds/trigger.
- cache the `glaredb` binary explicitly (saves after the build task, restored before any task that runs SLTs or needs glaredb (pytests) keyed off the current _run_ id. 
- save a second cache of `target` after the build task, restored before any task that needs to do on a build. key'd off the current commit. 

Then we restore exactly what we need before the task that it runs. 

This means we end up caching many more build-specific things and so the cache accelerates what we're doing in the task. and we don't end up thrashing the cache and rebuilding with different flags as much (e.g. unittests shouldn't thrash the build cache) 

The "workspace cache" is maybe of questionable effect, but it seems pretty harmless.